processing/src/main/java/io/druid/query/QueryDataSource.java

```diff
@@ -40,7 +42,10 @@ public QueryDataSource(@JsonProperty("query") Query query)
   @Override
   public List<String> getNames()
   {
-    return query.getDataSource().getNames();
+    final Iterable<DataSourceWithSegmentSpec> sourceSpecs = query.getDataSources();
+    return StreamSupport.stream(sourceSpecs.spliterator(), false)
+                 .flatMap(spec -> spec.getDataSource().getNames().stream())
+                 .collect(Collectors.toList());
```

Comment:
Since every data source in druid has a unique name, there is only one case if there are duplicated names in this stream. That is, the same data source appears multiple times in the  ```query```'s data sources like in self-join queries. In this case, that data source's name should be included multiple times in the result.